java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -680,3 +678,18 @@ private MethodAccess callReturningSameType(Expr ref) {
   ref = result.getQualifier() and
   result.getMethod().getReturnType() = ref.getType()
 }
+
+private SrcRefType entrypointType() {
+  result =
+    any(RemoteFlowSource s | s instanceof DataFlow::ExplicitParameterNode)
+        .getType()
+        .(RefType)
+        .getSourceDeclaration() or
+  result.getASourceSupertype() = entrypointType() or
```

Comment:
Hmm, actually when looking at the test, I'm starting to think we shouldn't have this line. A field can easily be of type `Object` (or some type parameter with no bound), and then this line will yield all types, which isn't really what we want.
Perhaps it would be better to implement this by collecting fields rather than types?

Reply:
I've been thinking about this and testing different variations, but I can't wrap my head around how that could be done differently. To collect fields recursively, we need to traverse the types anyway, right? And those fields could be supertypes of other types which in turn could have additional fields.

🤔  I'm starting to think that we may simply remove the supertype part. Just consider the declared types and their fields, recursively, but don't take potential subclasses into account, to avoid adding something like `Object` to the mix and thus creating taint steps for _every_ field of _every_ type. We'll probably miss some taint steps with this approach, though.

In any case, if I misunderstood what you meant by "collecting fields rather than types", please let me know.

Reply:
Removing the subtypes of fields but not of the source classes themselves seems to produce pretty sensible results, see 59d9c4f